File: week3/236_lowestCommonAncestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode res;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        this.res = null;
        boolean r = dfs(root, p, q);
        return res;
    }

    private boolean dfs(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return false;
        }
        boolean leftSon = dfs(root.left, p, q);
        boolean rightSon = dfs(root.right, p, q);
        if ((leftSon && rightSon && (root == p || root == q) ) || leftSon || rightSon) {
            this.res = root;
        }
        return leftSon || rightSon || (root.val == p.val || root.val == q.val);
    }

}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution2 {

    TreeNode ans = null;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root, p, q);
        return this.ans;
    }

    private boolean dfs(TreeNode cur, TreeNode p, TreeNode q) {
        if (cur == null) {
            return false;
        }
        int left = dfs(cur.left, p, q) ? 1: 0;
        int right = dfs(cur.right, p, q) ? 1: 0;
        int mid = (cur == p || cur == q) ? 1: 0;

        if (mid + left + right >= 2) {
            this.ans = cur;
        }
        return (mid + left + right > 0);
    }
}

File: week3/236_lowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:

    def __init__(self):
        self.ans = None
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        def dfs(cur):
            if not cur:
                return False

            left = dfs(cur.left)
            right = dfs(cur.right)
            mid = cur == p or cur == q

            if mid + left + right >= 2:
                self.ans = cur
            return mid or left or right

        dfs(root)
        return self.ans

#     Time: O(N)
#     Space:O(N)

